### PR TITLE
[SPARK-10562][SQL] support mixed case partitionBy column names for tables stored in metastore

### PR DESCRIPTION
**Pull request by cloud-fan**
https://issues.apache.org/jira/browse/SPARK-10562

### CODE REVIEW

**Comment by cloud-fan** [comment]
The root case is: hive metastore is case insensitive and also not case preserving.

We can lower case the field name that is used as partitioned column, to make it consistent with hive metastore.
Or we can use table properties to store the partitioned column names, do not trust hive metastore.

I choose the second approach in this PR, what do you think? @marmbrus @yhuai

**Comment by yhuai** [comment]
If we use the second approach, can user query this table from hive side? Let's assume that the user can use msck command to recover partitions (https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-RecoverPartitions(MSCKREPAIRTABLE)). My main question is that if we preserve the partition column names' case sensitivity in fs paths, will hive be able to read the data? Or Hive will use all lower-case paths and then fails to read those partitions?

**Comment by liancheng** [comment]
Did the following experiments against Hive 0.13.1, seems that Hive doesn't accept partition directory with uppercase letters in partition column names.

First create a simple partitioned table (the `src` table is the one we commonly used in Hive tests):

```sql
DROP TABLE IF EXISTS t;
CREATE TABLE t (key INT, value STRING) PARTITIONED BY (part INT);
INSERT OVERWRITE TABLE t PARTITION (part = 0) SELECT * FROM src;
INSERT OVERWRITE TABLE t PARTITION (part = 1) SELECT * FROM src;
```

Then manually rename the `part=0` partition directory to `ParT=0`. It turned out that Hive doesn't recognize this partition anymore, and `SELECT * FROM part = 0` gave empty result. `MSCK REPAIR TABLE t` doesn't work either.

**Comment by liancheng** [comment]
Another issue I noticed while experimenting was that, if field names containing uppercase letters in a Parquet file, Hive doesn't recognize it either. First create a Parquet table from Spark SQL side and save it as a Hive table:

```scala
val df = sqlContext.range(3).coalesce(1).selectExpr("id AS VaLuE")
df.write.mode("overwrite").saveAsTable("parq")
```

Although `SHOW TABLES` and `DESC parq` shows that Hive does recognize table `parq`, `SELECT * FROM parq` returns only `NULL` because the field name contains uppercase letters. Changing `VaLuE` to `value`, then everything is fine.

Maybe we should lowercase all field names when persisting a DataFrame to metastore in Hive compatible mode.

**Comment by cloud-fan** [comment]
When we write hive table in non-Hive-compatible mode, we [ignore the schema info](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala#L254-L263).
It looks to me that there is no workaround to read spark sql partitioned table at hive side anyway. So I think we don't need to care about Hive compablity when fix 10562, and makes the second approach a better choice.

**Comment by liancheng** [comment]
I'd also vote for the 2nd approach. PR #9251 doesn't exactly fix "mix case" `partitionBy`, but simply lowercases all partition columns. Two drawbacks of this approaches are:

1. Usually Spark SQL always tries to be case preserving, so that we can adapt to either case sensitive or case insensitive environments. But PR #9251 erases case information
2. This can be confusing, because even when users are working under case sensitive configuration, partition column names are still erased.

**Comment by yhuai** [comment]
@liancheng My concern is that if we do not lowercase those partition names, Hive will not be able to read those partition dirs.

**Comment by yhuai** [comment]
Had a discussion with @liancheng and @cloud-fan. We think this fix is better than #9251. When a data source table is partitioned, we will not save data in a Hive-compatible way. So, let's make sure we can read data back. Later, when we want to save partitioned data in a hive compatible way, we can discuss the right way for that in that JIRA (we may want to add a flag for it since the saved metadata may be quite different).

**Comment by yhuai** [comment]
Can we add an `assert` in `newHiveCompatibleMetastoreTable` to make it clear that we do not expect to pass a partitioned table to `newHiveCompatibleMetastoreTable`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -218,25 +233,19 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
...
       }
-      Seq.empty[HiveColumn]
+    } else {
+      // The table does not have a specified schema, which means that the schema will be inferred
+      // when we load the table. So, we are not expecting partition columns and we will discover
+      // partitions when we load the table. However, if there are specified partition columns,
+      // we simply ignore them and provide a warning message.
+      logWarning(
+        s"The schema and partitions of table $tableIdent will be inferred when it is loaded. " +
+          s"Specified partition columns (${partitionColumns.mkString(",")}) will be ignored.")
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Seems the condition of this branch is not equivalent to the original one? We may hit this branch if the table is not partitioned and there is a user-defined schema.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 2]
```diff
@@ -744,32 +744,6 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
...
-        s"Partitions columns stored in metastore $actualPartitionColumns is not the " +
-          s"partition columns defined by the saveAsTable operation $expectedPartitionColumns.")
-
-      // Check the content of the saved table.
-      checkAnswer(
-        table(tableName).select("c", "b", "d", "a"),
-        df.select("c", "b", "d", "a"))
-    }
-  }
-
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: Can we have a test to check the text contents of partition columns in the table properties?

**Comment by yhuai** [comment]
Overall loos good! Left two minor comments.

**Comment by yhuai** [comment]
LGTM. Merging to master.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 1 above)
Comment (yang0228) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Can you explain the change at line 293 "partitionColumns = Nil"? 
When running "df.write.partitionBy("X").mode(SaveMode.Overwrite).saveAsTable(tableName) "(df generated with HiveContext) in spark 1.6, what I get in hive is a not-partitioned table. I've seen this line was "partitionColumns = partitionColumns" in spark1.5.2, which assigns the partition column.

Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: line 293 "partitionColumns = Nil" is after an asset: `assert(partitionColumns.isEmpty)`.
We can not write out partitioned table in a hive compatible way, so I think this PR is not the root cause that break your code.
But it's surprising that the written table is partitioned at hive side in 1.5.2, can you double check it?

Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: The root cause is line 270. Even in non-hive compatible mode, we still keep partition info in hive metadata before. This PR changed it and store partition info as properties.

We can't write out partitioned table in a hive-compatible way, so I think it's a safe change, you can't query that table at hive side anyway.

**Pull request closed without merging**